### ISSUE TITLE
BitBucket Private Repositories

### ISSUE THREAD

**Issue by PthPndr** [yafred/wp-alt-repositories#2]
I recently found out about this plugin and wanted to give it a try. Should it work with BitBucket private repositories? 

I got it to work with public repositories but I get the following errors when attempting private repositories:

ERROR: Forbidden (you must use a valid authentication)

and

Notice: Trying to get property of non-object in \wp-content\plugins\wp-alt-repositories\WpAltRepo\Entities\BitbucketConnector.php on line 118

It's the same authentication I use with afragen's GitHub-Updater, an account that has Read Only access to the repository.

**Comment by yafred** [comment]
Yes it should ...

Go to Alt. Repo. > Connectors ... Add New

Select Bitbucket for the repository type
Set your user and password

Go to Alt. Repo. > Plugins (or Themes which ever your have created)
Edit the entry for your private repository 
Select your new connector in the Repository connector drop-down

Click on update plugin (or theme) definition 

Does it help ?


Note: the notice you get should not happen ... I'll correct it ... thanks for reporting

**Comment by PthPndr** [comment]
Yeah, that's what I tried but I am still getting the "ERROR: Forbidden (you must use a valid authentication)" error when I try to install a new plugin.

I tried using two different BitBucket accounts but both had the same result.

**Comment by yafred** [comment]
I'm afraid you're right ... This does work anymore.

I'll try and find what have changed

**Comment by yafred** [comment]
I corrected the bug on the dev branch

You can switch to the dev branch following these steps:
Go to Alt. Repo. > Plugins
Edit yafred/wp-alt-repositories plugin definition
*Repository state:* Install the most recent state (i.e. commit) of a branch.
*Branch/Tag/Commit:* dev
Click on Update plugin definition
Click on update now !

**Comment by PthPndr** [comment]
I updated the plugin and tried it again with no luck. I also tried wiping everything and doing it again but the same result. It's a slightly different error message this time:

ERROR: Unauthorized (you must use a valid authentication)

**Comment by yafred** [comment]
Please, can you check you have access to the repo.

Go to http://restbrowser.bitbucket.org/
GET https://bitbucket.org/api/2.0/repositories/{user}/{repo}/commits/master
using basic auth (put the user / password you tried with this plugin)

Thanks

**Comment by PthPndr** [comment]
Not sure if you're looking for something in particular but here's a snippet of the result.

HTTP/1.1 200 OK
ETag: "0bf3e22800d0032db8fcd94775eb8cbb"
X-Render-Time: 0.08837890625
Content-Length: 6079
X-Served-By: app19
Last-Modified: Tue, 06 Oct 2015 15:48:03 GMT
X-Version: d4994ede4cf5
Connection: keep-alive
X-Request-Count: 197
Server: nginx/1.6.2
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000
Date: Fri, 16 Oct 2015 11:32:04 GMT
Vary: Authorization, Cookie
X-Cache-Status: BYPASS
X-Static-Version: d4994ede4cf5
Content-Type: application/json; charset=utf-8
X-Accepted-OAuth-Scopes: repository

**Comment by yafred** [comment]
200 OK means you have access.
The plugin runs the same request, but it gets ERROR: Unauthorized (you must use a valid authentication)

As the dev branch works fine for me (accessing a private bitbucket repository), there are a few things you could double check:

1. Are you running the right code ?
Please check file WpAltRepo/Entities/BitbucketConnector.php 
You should find this code 3 times

> add_filter( 'http_request_args', array($this, 'basicAuth'), 10, 2 );

2. Check again you have selected a connector that uses the user / password you used in the bitbucket rest browser (sorry, I'm sure you already did that)

3. If you are using other plugins can you disable them for a test.
I wonder if your credentials are overwritten by somebody else using

> add_filter( 'http_request_args', array($this, 'basicAuth'), 10, 2 );

**Comment by PthPndr** [comment]
I deactivated all plugins except yours and the result was still the same. I'm running this on a local VM for testing so I built a regular server from scratch and installed it on there. It does appear to work correctly on there. No authentication errors.

Quick question, how long does it usually take to notice an update is needed?
And if you force an update from the WordPress console will it also trigger a check?

**Comment by yafred** [comment]
> Quick question, how long does it usually take to notice an update is needed?

12 hours maximum. It is hard coded [here] (https://github.com/yafred/wp-alt-repositories/blob/610fbba3433b9e318c725863bd7bbbc869a56ac3/WpAltRepo/Main.php#L13). We could think of a filter for you to configure it.

> And if you force an update from the WordPress console will it also trigger a check?

No. At the moment, you can only force the check in the plugin or theme edit panel. So, one repository at a time. I'll think of more options for this topic.

Some bad news: I found another bug (bitbucket API seems to have changed since I last used this piece of code). When you try to update a plugin hosted in a private Bitbucket repository, you should get an error message

> The package could not be installed. PCLZIP_ERR_BAD_FORMAT (-10) : Unable to find End of Central Dir Record signature

I'll dig into it
